src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -343,6 +348,20 @@ public synchronized void subscribe(SubscribeRequest request)
                 .findAny();
     }
 
+    @SuppressFBWarnings("JLM_JSR166_UTILCONCURRENT_MONITORENTER")
+    private void triggerSpooler() {
+        // Do not synchronize on MqttClient because that causes a dead lock
+        synchronized (spoolingFuture) {
+            if (spoolingFuture.get() == null || spoolingFuture.get().isDone() || spoolingFuture.get().isCancelled()) {
```

Comment:
Why check `spoolingFuture.get().isCancelled()` and `spoolingFuture.get().isDone()`? `isCancelled()` and `isDone()` only returns true is `spoolingFuture` is cancelled or done, which can only happen when spooler is shutdown (the runSpooler task thread is interrupted). The two condition checks seem unnecessary. This also introduces race condition between `triggerSpooler()` method and `close()` method.

Reply:
If the spooler dies or quits because the network is out (the thread quits when mqtt is offline)

Reply:
Alright. I missed the first break statement. Then we need to think about how to address the race condition.

Reply:
@awszztt just remove the `isCancelled` check and that will resolve the race.

Reply:
yes.

Reply:
If a future is cancelled, isDone() is also true. So I'm not sure if that is sufficient.